File: src/main/java/com/timeback/device/os/SystemTimeSource.java
package com.timeback.device.os;

import com.timeback.device.contract.TimeRange;
import com.timeback.device.contract.TimeSource;

import java.time.Clock;
import java.time.ZoneId;
import java.util.List;
import java.util.function.Supplier;

public final class SystemTimeSource implements TimeSource {
    private final Clock clock;
    private final Supplier<ZoneId> systemZone;

    public SystemTimeSource() {
        this(Clock.systemUTC(), ZoneId::systemDefault);
    }

    public SystemTimeSource(Clock clock, Supplier<ZoneId> systemZone) {
        this.clock = clock;
        this.systemZone = systemZone;
    }

    @Override
    public long nowMillis() {
        return clock.millis();
    }

    @Override
    public ZoneId zoneId() {
        return systemZone.get();
    }

    @Override
    public List<Long> localMidnightBoundaries(TimeRange range) {
        return TimeBoundaryCalculator.localMidnightBoundaries(range, zoneId());
    }
}

File: src/main/java/com/timeback/device/os/SystemTimeSource.kt
package com.timeback.device.os

import com.timeback.device.contract.TimeRange
import com.timeback.device.contract.TimeSource
import java.time.Clock
import java.time.Instant
import java.time.ZoneId

class SystemTimeSource(
    private val clock: Clock = Clock.systemUTC(),
    private val systemZone: () -> ZoneId = { ZoneId.systemDefault() },
) : TimeSource {
    override fun nowMillis(): Long = clock.millis()

    override fun zoneId(): ZoneId = systemZone()

    override fun localMidnightBoundaries(range: TimeRange): List<Long> =
        TimeBoundaryCalculator.localMidnightBoundaries(range, zoneId())
}

object TimeBoundaryCalculator {
    fun localMidnightBoundaries(range: TimeRange, zoneId: ZoneId): List<Long> {
        var nextDate = Instant.ofEpochMilli(range.startAtMillis)
            .atZone(zoneId)
            .toLocalDate()
            .plusDays(1)
        val boundaries = mutableListOf<Long>()
        while (true) {
            val boundary = nextDate.atStartOfDay(zoneId).toInstant().toEpochMilli()
            if (boundary >= range.endAtMillis) break
            if (boundary > range.startAtMillis) boundaries += boundary
            nextDate = nextDate.plusDays(1)
        }
        return boundaries
    }
}
